src/main/kotlin/org/rust/ide/intentions/ChangeVisibilityIntention.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.cargo.project.workspace.PackageOrigin
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+class ChangeVisibilityIntention : RsElementBaseIntentionAction<ChangeVisibilityIntention.Context>() {
+    override fun getFamilyName(): String = "Change item visibility"
+
+    data class Context(val element: RsVisibilityOwner)
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val visibleElement = element.ancestorStrict<RsVisibilityOwner>() ?: return null
+        if (!isAvailable(visibleElement)) return null
+
+        val isVisible = visibleElement.isPublic
+        val name = (visibleElement as? RsNamedElement)?.name?.let { " `$it`" } ?: ""
+        val newVisibility = if (isVisible) "private" else "public"
+
+        text = "Make$name $newVisibility"
+
+        return Context(visibleElement)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        if (ctx.element.isPublic) {
+            makePrivate(ctx.element)
+        } else {
+            makePublic(ctx.element, true)
+        }
+    }
+
+    companion object {
+        fun isAvailable(visible: RsVisible): Boolean =
+            visible !is RsEnumVariant
+                && visible.parent.parent !is RsTraitItem
+                && visible.containingCrate?.origin == PackageOrigin.WORKSPACE
+
+        fun makePrivate(element: RsVisibilityOwner) {
+            element.vis?.delete()
+        }
+        fun makePublic(element: RsVisibilityOwner, withinOneCrate: Boolean) {
+            val project = element.project
+
+            var anchor = getAnchor(element)
+            // Check if there are any elements between `pub` and type keyword like
+            // `extern "C"`, `unsafe`, `const`, `async` etc.
+            // If prevNonCommentSibling is an attribute or null, then
+            // there are no extra words, listed above, and we can
+            // freely insert `pub` element right before keyword element.
+            // Otherwise, there are some of these words and we roll anchor back to
+            // insert `pub` element before them.
+            while (true) {
+                val prevNonCommentSibling = anchor.getPrevNonCommentSibling()
+                if (!(prevNonCommentSibling is RsOuterAttr || prevNonCommentSibling == null)) {
+                    anchor = prevNonCommentSibling
+                } else {
+                    break
+                }
+            }
+            if (!withinOneCrate) {
```

Comment:
I think, inverting this condition could make the meaning of `withinOneCrate` more clear :)